### PR TITLE
LEAN-2992

### PR DESCRIPTION
**Pull request by mnatsheh**

### CODE REVIEW

**Comment by mbartenev-atypon** [comment]
I thought the plan was to move menus into the article editor or did I miss something?

**Comment by mnatsheh** [comment]
> I thought the plan was to move menus into the article editor or did I miss something?

Since the body-editor needed the types I thought I would just put it here. An alternative is to put the types in body-editor and the business logic in article-editor (which is basically what I did for the toolbar), so I'm happy to go with that route if it's preferred.

**Comment by mbartenev-atypon** [comment]
> > I thought the plan was to move menus into the article editor or did I miss something?
> 
> Since the body-editor needed the types I thought I would just put it here. An alternative is to put the types in body-editor and the business logic in article-editor (which is basically what I did for the toolbar), so I'm happy to go with that route if it's preferred.

I would probably prefer to minimise the usage of prosemirror components in the styleguide and semnatically make it clear that menus shouldn't be used anywhere else so if that's not too much trouble I'd suggest to move them to the article-editor. But this is not critical.

**Review by mbartenev-atypon — APPROVED** [review]

File: src/components/projects/ManuscriptMenus.tsx [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    })
+    .filter(Boolean) as (MenuSpec | MenuSeparator)[]
+}
+
+export const ManuscriptMenus: React.FC<ApplicationMenusProps> = ({
+  editor,
+}) => {
+  const can = usePermissions()
+
+  const specs = processMenu(getEditorMenus(editor), (spec: MenuSpec) => {
```
Comment (mbartenev-atypon) [inline src/components/projects/ManuscriptMenus.tsx:72]: This calculations are complex enough to be wrapped with useMemo.

File: src/components/projects/ManuscriptMenus.tsx [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+} from '@manuscripts/style-guide'
+import React from 'react'
+
+import config from '../../config'
+
+interface ApplicationMenusProps {
+  editor: ReturnType<typeof useEditor>
+}
+
+const processMenu = (
```
Comment (mbartenev-atypon) [inline src/components/projects/ManuscriptMenus.tsx:30]: This seems to be a very generic name. Perhaps something like a permissionFilter can be more descriptive? What do you think?

File: src/components/projects/ManuscriptMenus.tsx [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        if (processed.submenu) {
+          processed.submenu = processSubmenu(processed.submenu, fn)
+        }
+        return processed
+      }
+    })
+    .filter(Boolean) as MenuSpec[]
+}
+
+const processSubmenu = (
```
Comment (mbartenev-atypon) [inline src/components/projects/ManuscriptMenus.tsx:47]: Seems like the difference between processMenu and processSubmenu are minimal and these 2 functions can be combined into one

File: src/components/projects/OrderedListToolbarItem.tsx [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+  )
+}
+
+/**
+ *  Will use node depth as an indicator for the list level, and `sharedDepth` will
+ *  give us all depths for the selection so we can find the parent list
+ *  After that will look at the nodes in the parent list and just update
+ *  the list that had the same level to the target level user has selected
+ */
+export const updateListStyle = (
```
Comment (mbartenev-atypon) [inline src/components/projects/OrderedListToolbarItem.tsx:189]: In my opinion all this functionality should be in the body-editor and just imported to be used in the tsx components here.

File: src/components/projects/OrderedListToolbarItem.tsx [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+      return (
+        node.type === schema.nodes.ordered_list ||
+        node.type === schema.nodes.bullet_list
+      )
+    }) + 1
+  const parentListNode = $from.node(parentListDepth)
+  const parentListPos = $from.before(parentListDepth)
+
+  if (parentListDepth === targetLevel) {
+    replaceToOrderList(
```
Comment (mbartenev-atypon) [inline src/components/projects/OrderedListToolbarItem.tsx:211]: replaceToOrderList is one argument away from being a pure function declared out this function scope. I know it's copied so this is optional for this PR.

**Review by mbartenev-atypon — CHANGES_REQUESTED** [review]

**Comment by mbartenev-atypon** [comment]
In general I'm ok with this approach but it seems that the menus functionality is now spread across 3 component. Some parts are in the style-guide, some in the article-editor and some are in the style-guide. I'm not sure how beneficial it is since menus and not shared between multiple components and the usage of the stories does not appear to be very useful for us anyway. But we can get back to this sometime later after this is merged.

**Comment by mbartenev-atypon** [comment]
> @mbartenev-atypon, there is no prosemirror usage in the menu logic that was moved to style-guide. I thought about it, and having stories in style-guide is why I would like to have that logic there (so it can change without having the run a full app to try it).

ah, yes I see now. Thanks

File: src/components/projects/OrderedListToolbarItem.tsx (hunk 4 above)
Comment (mnatsheh) [reply src/components/projects/OrderedListToolbarItem.tsx]: Agreed. I've moved the logic to body-editor and simplified it as it was convoluted for what it does.

File: src/components/projects/ManuscriptMenus.tsx (hunk 2 above)
Comment (mnatsheh) [reply src/components/projects/ManuscriptMenus.tsx]: Changed.

File: src/components/projects/ManuscriptMenus.tsx (hunk 3 above)
Comment (mnatsheh) [reply src/components/projects/ManuscriptMenus.tsx]: I've refactored the code to eliminate the duplication.

File: src/components/projects/ManuscriptMenus.tsx (hunk 1 above)
Comment (mnatsheh) [reply src/components/projects/ManuscriptMenus.tsx]: Done.

**Review by mbartenev-atypon — APPROVED** [review]

**Pull request merged**